### PR TITLE
Merge shared semaphores routines from OpenJ9

### PR DESCRIPTION
**Pull request by mthom**
This commit is in support of the refactored shared cache port.

Issue: #4610

Signed-off-by: Mark Thom <mark.thom@unb.ca>

### CODE REVIEW

**Comment by rwy0717** [comment]
There's quite a few whitespace errors in here. Indentation should be done with tabs, and alignment should be done with spaces.

File: include_core/omrport.h [hunk 1]
```diff
@@ -284,6 +284,55 @@
...
+#define OMRSH_SYSV_REGULAR_CONTROL_FILE 0
+#define OMRSH_SYSV_OLDER_CONTROL_FILE 1
+#define OMRSH_SYSV_OLDER_EMPTY_CONTROL_FILE 2
+
+/* Flags passed to "flag" argument of omrshsem_deprecated_open(). */
+#define OMRSHSEM_NO_FLAGS		0x0
+#define OMRSHSEM_OPEN_FOR_STATS		0x1
+#define OMRSHSEM_OPEN_FOR_DESTROY	0x2
+#define OMRSHSEM_OPEN_DO_NOT_CREATE     0x4
+
```
Comment (rwy0717) [inline include_core/omrport.h]: For example, these constants should be aligned with spaces, or not aligned at all.

File: include_core/omrport.h [hunk 2]
```diff
@@ -297,6 +346,18 @@ typedef struct J9Permission {
...
+        uintptr_t semid;
+        uintptr_t ouid;
+        uintptr_t ogid;
+        uintptr_t cuid;
+	uintptr_t cgid;
+	int64_t lastOpTime;
+	int64_t lastChangeTime;
+	int32_t nsems;
+        J9Permission perm;
+} OMRPortShsemStatistic;
```
Comment (rwy0717) [inline include_core/omrport.h]: Another example: this struct is using a mix of tabs and spaces for indentation, when it should be all tabs.

**Comment by rwy0717** [comment]
As well, the copyrights should be updated to 2019 or 2020.

**Comment by 0xdaryl** [comment]
Just so we get this out of the way now, please remove the trailing period from the commit titles of this and all your other recent PRs per the requirements of the Contribution Guidelines.  Thanks!  :-)

File: include_core/omrport.h (hunk 1 above)
Comment (DmonteDD) [reply include_core/omrport.h:337]: Thank you for your comment. Modified in latest commit.

File: include_core/omrport.h (hunk 2 above)
Comment (DmonteDD) [reply include_core/omrport.h:361]: Thank you for your comment. Modified in latest commit.

**Comment by DmonteDD** [comment]
Dear reviewers, @rwy0717, @youngar and @mstoodle . I have updated this PR asper suggestions in comments. Can you please suggest next round of improvements, or are we good to procced?

File: port/common/omrport.c [hunk 3]
```diff
@@ -259,6 +259,31 @@ static OMRPortLibrary MainPortLibraryTable = {
...
+	omrshsem_startup, /* shsem_startup */
+	omrshsem_shutdown, /* shsem_shutdown */
+	omrshsem_open, /* shsem_open */
+	omrshsem_post, /* shsem_post */
+	omrshsem_wait, /* shsem_wait */
+	omrshsem_getVal, /* shsem_getVal */
+	omrshsem_setVal, /* shsem_setVal */
+	omrshsem_close, /* shsem_close */
+	omrshsem_destroy, /* shsem_destroy */
+	omrshsem_deprecated_startup, /* shsem_deprecated_startup */
```
Comment (youngar) [inline port/common/omrport.c:273]: Do we need to move the deprecated functions?

File: port/common/omrshsem_deprecated.c [hunk 4]
```diff
@@ -0,0 +1,1276 @@
...
+		Trc_PRT_shsem_omrshsem_openDeprecated_Message("Error: could not lock semaphore control file.");
+		goto error;
+	}
+
+	if (cacheFileType == OMRSH_SYSV_OLDER_EMPTY_CONTROL_FILE) {
+		int semid = -1;
+		int projid = 0xad;
+		key_t fkey;
+		int semflags_open = groupPerm ? OMRSHSEM_SEMFLAGS_GROUP_OPEN : OMRSHSEM_SEMFLAGS_OPEN;
+//		intptr_t rc;
```
Comment (youngar) [inline port/common/omrshsem_deprecated.c:128]:
```suggestion

```

File: port/sysvipc/omrshsem.h [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+	int32_t semsetSize;
+} omrshsem_baseFileFormat;
+
+typedef struct omrshsem_handle {
+	int32_t semid;
+	int32_t nsems;
+	char* baseFile;
+	int64_t timestamp;
+ 	uint32_t deleteBasefile : 1; /* delete the base file (used to generate the semaphore key) when destroying the semaphore */
+	/* int32_t baseFilefd deleted: it is never assigned or used. */
```
Comment (youngar) [inline port/sysvipc/omrshsem.h:49]:
```suggestion
```

File: port/sysvipc/omrshsem.h [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+	int32_t creator_pid;
+	int32_t semsetSize;
+} omrshsem_baseFileFormat;
+
+typedef struct omrshsem_handle {
+	int32_t semid;
+	int32_t nsems;
+	char* baseFile;
+	int64_t timestamp;
+ 	uint32_t deleteBasefile : 1; /* delete the base file (used to generate the semaphore key) when destroying the semaphore */
```
Comment (youngar) [inline port/sysvipc/omrshsem.h:48]:
```suggestion
	uint32_t deleteBasefile : 1; /* delete the base file (used to generate the semaphore key) when destroying the semaphore */
```

**Comment by youngar** [comment]
For `port/sysvipc/omrSysvIPCWrappers.c`
1. I think most (all?) port library files use lower case letters in file names
2. Would it make more sense to put these files into `port/unix` instead of creating a new sysvipc folder?

File: port/common/omrport.c (hunk 3 above)
Comment (DmonteDD) [reply port/common/omrport.c:273]: Yes, these functions are necessary as they will be consumed by other shared cache files which are included in other PR's.

Comment (youngar) [reply port/common/omrport.c:273]: Thanks for the explanation!

**Comment by DmonteDD** [comment]
> For `port/sysvipc/omrSysvIPCWrappers.c`
> 
> 1. I think most (all?) port library files use lower case letters in file names
> 2. Would it make more sense to put these files into `port/unix` instead of creating a new sysvipc folder?

Thanks for the comment! 

1. The name of the corresponding file in OpenJ9 port is j9SysvIPCWrappers.c. Hence, it was kept similar. I guess changing it wont be an issue. I will change it to omrsysv_ipcwrappers.c. 

2. sysvipc folder in OMR port is created as OpenJ9 port follows a similar structure. It places all SysV-style shared memory and semaphore routines that are required by the shared classes cache in the sysvipc folder. Hence, the same structure is followed in OMR. Yes, it makes sense to move those those files in unix folder as it also contains SysV routines.

I will incorporate all your suggestion in the next push!

**Comment by youngar** [comment]
@genie-omr build all

**Comment by youngar** [comment]
Thank you @DmonteDD and @mthom!

**Comment by DmonteDD** [comment]
> Thank you @DmonteDD and @mthom!

You're welcome! 
I would like to thank all reviewers for their comments and suggestions!

**Pull request merged**